Yosemite/Yosemite/Stores/OrderStore.swift

```diff
@@ -77,6 +83,95 @@ private extension OrderStore {
+    ///
+    /// If `deleteAllBeforeSaving` is true, all the orders will be deleted before saving any newly
+    /// fetched `Order`. The deletion only happens once, regardless of the which fetch request
+    /// finishes first.
+    ///
+    /// The orders will only be deleted if one of the executed `GET` requests succeed.
+    ///
+    /// - Parameter statusKey The status to use for the filtered list. If this is not provided,
+    ///                       only the all orders list will be fetched. See `OrderStatusEnum`
+    ///                       for possible values.
+    ///
+    func fetchFilteredAndAllOrders(siteID: Int64,
+                                   statusKey: String?,
+                                   deleteAllBeforeSaving: Bool,
+                                   pageSize: Int,
+                                   onCompletion: @escaping (Error?) -> Void) {
+
+        let remote = OrdersRemote(network: network)
+        let pageNumber = OrdersRemote.Defaults.pageNumber
+
+        // Synchronous variables.
+        //
+        // The variables `fetchErrors` and `hasDeletedAllOrders` should only be accessed
+        // **inside** the `serialQueue` (e.g. `serialQueue.async()`). The only exception is in
+        // the `group.notify()` call below which only _reads_ `fetchErrors` and all the _writes_
+        // have finished.
+        var fetchErrors = [Error]()
+        var hasDeletedAllOrders = false
+        let serialQueue = DispatchQueue(label: "orders_sync", qos: .userInitiated)
+
+        // Delete all the orders if we haven't yet.
+        // This should only be called inside a `serialQueue` block.
+        let deleteAllOrdersOnce = {
+            guard hasDeletedAllOrders == false else {
+                return
+            }
+
+            // Use the main thread because `resetStoredOrders` uses `viewStorage`.
+            DispatchQueue.main.sync { [weak self] in
+                self?.resetStoredOrders { }
+            }
+
+            hasDeletedAllOrders = true
+        }
+
+        // The handler for both dual fetch requests.
+        let loadAllOrders: (String?, @escaping (() -> Void)) -> Void = { statusKey, completion in
+            remote.loadAllOrders(for: siteID, statusKey: statusKey, pageNumber: pageNumber, pageSize: pageSize) { orders, error in
+                serialQueue.async { [weak self] in
+                    guard let self = self else {
+                        completion()
+                        return
+                    }
+
+                    if let orders = orders {
+                        if deleteAllBeforeSaving {
+                            deleteAllOrdersOnce()
+                        }
+
+                        self.upsertStoredOrdersInBackground(readOnlyOrders: orders, onCompletion: completion)
+                    } else if let error = error {
+                        fetchErrors.append(error)
+                        completion()
+                    }
```

Comment:
just wondering, is it ever possible the `orders` and `error` are both nil? not sure if it's worth handling the `else` case just in any case?

Reply:
I don't think so. I think this should be an `Either<[Order], Error]` or `Result<[Order], Error>`. But we don't have that construct yet. So in here, I'm using `if let` as a shortcut for _optional checking_.

Reply:
> not sure if it's worth handling the else case just in any case?

Added thorough `completion` call in f12cb3a010ed1393d620a49bd989ad6a6890d6af. Thanks for catching that!